---
title: "parse input files"
output: html_document
date: "2025-07-10"
---

Here we are reading all results and mapfiles. creating required input files
for functions and analyses

```{r}

library(tidyverse)
library(ape) # for reading/writing tree files
#load functions
source('./functions.R')

```

## Mapfile
###########################################################
One of parentA sample has a very low depth of sequencing
this must be removed from analyses
filter(Sample != "A1E10M")

```{r}

mapfile <- read.csv("../data/mapfile.txt", sep = "\t")
data <- read.csv("../data/read_counts.txt", 
                 header = F, sep = ":")
data %>%
  filter(grepl("_R1.fastq.gz", V1)) %>%
  mutate(Sample= gsub("_R1.fastq.gz", "", V1)) %>%
  rename(Reads=V2) %>%
  left_join(mapfile) %>%
  select(Sample, Reads, Group, Isolator, Cage, Category) %>%
  # sample with low depth:
  filter(Sample != "A1E10M") %>%
write.table("../data/clean/mapfile.txt", sep = "\t", 
            row.names = F, quote = F)

```

## MAG data

bins quality and taxonomy
```{r}

bin.qualy <- read.csv("../Results/Assembly/bins_quality.txt", sep = "\t")
bin.taxa <- read.csv("../Results/Assembly/bins_taxonomy.txt", sep = "\t")

bin.qualy %>%
  select(BinLab, Bin.Id,Completeness, Contamination) %>%
  left_join(bin.taxa %>%
            select(BinLab, Phylum:Species)) %>%
  # make a clean MAG label:
  mutate(family_lab= case_when(
  str_detect(Family, "\\d") ~ paste0("o_", Order),
  TRUE ~ paste0(Family))) %>%
  mutate(genus_lab = case_when(
  Genus == "" ~ paste0("f_", family_lab),
  str_detect(Genus, "\\d") ~ paste0("f_", family_lab),
  TRUE ~ paste0("g_", Genus))) %>%
  mutate(species_lab= case_when(
    Species == "" ~ paste0(genus_lab),
    str_detect(Species, "CAG") ~ paste0(genus_lab),
    str_detect(Species, "UBA") ~ paste0(genus_lab),
    str_detect(Species, "COE") ~ paste0(genus_lab),
    str_detect(Species, "RC9") ~ paste0(genus_lab),
    TRUE ~ Species)) %>%
  mutate(label = paste(species_lab, 
                        gsub("_", "", Bin.Id),
                        sep = "_")) %>%
  select(-Bin.Id, -family_lab, -genus_lab, -species_lab) %>%
write.table("../data/clean/binQT.txt", sep = "\t", 
            row.names = F, quote = F)

```

bins 1x coverage data across all samples. this is rarefied reads to lowest number
```{r}

contigs <- read.csv("../Results/Assembly/contig_in_bins_anvi.txt", sep = "\t",
                    header = F)
colnames(contigs) <- c("contig", "BinLab")

cover_parser(path = "../Results/p_map_subset/contigs_1k",
             patt= "_perfect.cover") %>%
  # adding bin info to contigs
  mutate(Don= gsub("_pool__.*", "", contig)) %>%
  left_join(contigs) %>%
  # define contigs outside of bins
  mutate(BinLab= case_when(is.na(BinLab) ~ paste(Don,"UnBin", sep = "_"),
                           TRUE ~ paste(BinLab))) %>%
  # Convert columns with numeric characters to numeric type
  mutate(across(everything(), ~ if (all(
    grepl("^\\s*-?\\d+(\\.\\d+)?\\s*$", .,
          perl = TRUE))) as.numeric(.) else .)) %>%
  group_by(Sample, BinLab) %>%
  summarise(
    bin_len = sum(endpos),
    bin_covbases = sum(covbases),
    bin_numreads = sum(numreads),
    bin_coverage = bin_covbases / bin_len * 100,
    .groups = "drop"
  ) %>%
  mutate(Sample = gsub("sub_", "", Sample)) %>%
  mutate(coverage = round(bin_coverage, digits = 2)) %>%
  rename(feature="BinLab") %>%
  select(Sample, feature, coverage) %>%
  # sample with low depth:
  filter(Sample != "A1E10M") %>%
write.table("../data/clean/binCover.txt", sep = "\t", 
            row.names = F, quote = F)
```

MAGs 1x coverage data across all samples.

```{r}

binQT <- read.delim("../data/clean/binQT.txt", 
           header = T, stringsAsFactors = F)

binCover <- read.delim("../data/clean/binCover.txt",
                       header = T, stringsAsFactors = F)

MAGid <- binQT %>%
  filter(Completeness >= 50 & Contamination <= 10) %>%
  pull(BinLab) %>% unique()

binCover %>%
  filter(feature %in% MAGid) %>%
write.table("../data/clean/MAGCover.txt", sep = "\t", 
            row.names = F, quote = F)

MAGid %>%
write.table("../data/clean/MAGids.txt", sep = "\t", 
            row.names = F, quote = F, col.names = F)

```

# MAGs detect in at least one sample of pups(AB/BA/ABAB):

```{r}

feature_c = "../data/clean/MAGCover.txt"
cutoff_pres = 75
method = "Assembly"
mapfile = "../data/clean/mapfile.txt"
# DonorA
donorA_feature <- DonFeature(feature_c = feature_c, donor = "DonA",
                           cutoff_pres = cutoff_pres, method = method)
finalA_identity <- identColoniz(donor_feature = donorA_feature, 
                               cutoff_pres = cutoff_pres,
                               mapfile = mapfile) 
finalA_identity %>% distinct(feature) %>%
  mutate(MAGs = gsub("DonA_bin_", "bin.", feature)) %>%
  mutate(MAGs = gsub("$", ".fa", MAGs)) %>%
  select(MAGs) %>%
  write.table("../data/clean/MAGids_DonA.txt", 
            row.names = F, quote = F, col.names = F)
# DonorB
donorB_feature <- DonFeature(feature_c = feature_c, donor = "DonB",
                           cutoff_pres = cutoff_pres, method = method)
finalB_identity <- identColoniz(donor_feature = donorB_feature, 
                               cutoff_pres = cutoff_pres,
                               mapfile = mapfile) 
finalB_identity %>% distinct(feature) %>%
  mutate(MAGs = gsub("DonB_bin_", "bin.", feature)) %>%
  mutate(MAGs = gsub("$", ".fa", MAGs)) %>%
  select(MAGs) %>%
  write.table("../data/clean/MAGids_DonB.txt", 
            row.names = F, quote = F, col.names = F)

```

MAG tree files, seperated for DonA and DonB
GTDB-tk was ran for MAGs and the alignment files were used to make a tree
using fasttree

```{r}

# Read a tree fix the label and keep it in clean data
DonA.tree <- read.tree("../Results/Assembly/DonA.bac120.user_msa.tree")
DonA.tree$tip.label <- gsub("bin.", "DonA_bin_", DonA.tree$tip.label)
write.tree(DonA.tree, "../data/clean/MAGs_DonA.tree")

DonB.tree <- read.tree("../Results/Assembly/DonB.bac120.user_msa.tree")
DonB.tree$tip.label <- gsub("bin.", "DonB_bin_", DonB.tree$tip.label)
write.tree(DonB.tree, "../data/clean/MAGs_DonB.tree")

```

## Gene data

gff file
```{r}

gff_cols <- c("seqname","source","feature","start","end","score","strand",
              "frame","attributes")
read.delim("../Results/p_map_subset/Donor_db.gff",
           header=F, comment.char="#", ) -> gff
colnames(gff) <- gff_cols

# removing sequence info from the end of the gff file
gff %>%
  mutate(ID= gsub(";.*", "", attributes)) %>%
  mutate(gene_id=gsub("ID=", "", ID)) %>%
  # annotated regions that have only contig info in attributes. 
  # I don't need them
  filter(feature!= "region") %>%
  filter(feature == "CDS")-> gff
write.table(gff, "../data/clean/gene_db.txt", sep = "\t", 
            row.names = F, quote = F)


```

Gene 1x coverage across all samples. reads rarefied to the lowest depth.
```{r}

gff %>%
  mutate(donor=gsub("_pool__.*", "", seqname)) %>%
  mutate(feature= paste(donor, gene_id, sep = "__")) %>%
  select(gene_id, feature) -> gene.ids

cover_parser(path = "../Results/p_map_subset/genes",
             patt= "_perfect.cover") %>%
  dplyr::rename(gene_id = "contig") %>%
  # genes must be >=100bp
  filter(endpos >= 100) %>%
  # genes that were identified as CDS in gff file
  left_join(gene.ids) %>%
  filter(!is.na(feature)) %>%
  select(Sample, feature, coverage) %>%
   # sample with low depth:
  filter(Sample != "A1E10M") %>%
write.table("../data/clean/geneCover.txt", sep = "\t", 
            row.names = F, quote = F)

```

## Species data

Relative abundance of species across sample using metaphlan4
database used:#mpa_vJan21_CHOCOPhlAnSGB_202103

```{r}

read.csv("../Results/Reads/merged_metaphlan4.txt", 
         skip = 1, sep = "\t") %>%
  filter(str_detect(clade_name, "\\|t__")) %>%
  gather(Sample, coverage, -clade_name) %>%
  mutate(Sample = gsub(".mp.profile", "", Sample)) %>%
  rename(feature = "clade_name") %>%
  select(Sample, feature, coverage) %>%
  # sample with low depth:
  filter(Sample != "A1E10M") %>%
write.table("../data/clean/speciesAbund.txt", sep = "\t", 
            row.names = F, quote = F)

```

## StrainPhlan data
a list of species (SGB) with min abundance in donor and parent samples for
StrainPhlan analyses
```{r}

spAbund <- read.delim("../data/clean/speciesAbund.txt",
                      header = T, stringsAsFactors = F)
map <- read.delim("../data/clean/mapfile.txt", header = T, stringsAsFactors = F)

# fist, it must be present in donor samples
cut_off=0.001
spAbund %>%
  left_join(map) %>%
  filter(Group %in% c("DonorA", "DonorB")) %>%
  filter(coverage >= cut_off) %>%
  select(Sample, feature) %>%
  distinct() %>% pull(feature) -> DonSp

spAbund %>%
  filter(feature %in% DonSp) %>%
  left_join(map) %>%
  filter(Group %in% c("ParentA", "ParentB")) %>%
  select(Sample, feature, coverage) %>%
  spread(Sample, coverage) %>%
# Filter features with coverage > cut_off in at least 5 Parent samples
  rowwise() %>%
  mutate(above_cutoff = sum(c_across(-feature) > cut_off, na.rm = TRUE)) %>%
  ungroup() %>%
  filter(above_cutoff >= 5) %>%
  select(-above_cutoff) %>%
  mutate(SGB = gsub(".*\\|t__", "t__", feature)) %>%
  select(SGB, feature) -> SGBs

write.table(SGBs, "../data/clean/DonAndParentSp_list.txt", sep = "\t",
            quote = F, row.names = F)
SGBs %>%
  pull(SGB) %>%
write.table("../data/clean/DonAndParentSp_SGBs.txt", sep = "\t",
            quote = F, row.names = F, col.names = F)


```

importing strainphlan distances for the SGB of interest.note that some of these
were failed because they were not present in enough samples, with enough 
coverage. we used the authors suggested cut-off for these values and lower than
the default settings. 

```{r}

map <- "../data/clean/mapfile.txt"
precom_thresh = "../Results/Reads/strainphlan_threshold/VallesColomerM_2022_Jan21_thresholds.tsv"
path = "../Results/Reads/strainphlan"
patt = "_nGD.tsv"

data.frame(marker.id = paste(dir(path, pattern = patt))) %>%
  mutate(file_contents = map(marker.id, ~ read_tsv(
      file.path(path, .),
      col_names = F))) %>%
  unnest() %>%
  mutate(marker.id = gsub(patt, "", marker.id)) -> pwdis

length(unique(pwdis$marker.id))

pwdis %>%
  # only samples compared to Wild116, and X608:
  filter(X2 %in% c("X608", "Wild116")) %>%
  mutate(pairwise= paste(X1, X2, sep = "-")) %>%
  mutate(
         pairwise= paste(marker.id, pairwise, sep = "-"),
         distance= X3,
         Sample= paste(X1),
         SGB= marker.id,
         Donor= paste(X2)) %>%
  select(pairwise, SGB, Sample, distance, Donor) -> don_pwdis

```


counts of colonized features at different levels. one table of all features.

```{r}
#Bins
CountFeature(feature_c = "../data/clean/binCover.txt",
            cutoff_pres = 75,
            method = "Assembly",
            mapfile = "../data/clean/mapfile.txt") %>%
  mutate(feature = paste("Bin")) -> count_bins
#MAGs
CountFeature(feature_c = "../data/clean/MAGCover.txt",
            cutoff_pres = 75,
            method = "Assembly",
            mapfile = "../data/clean/mapfile.txt") %>%
  mutate(feature = paste("MAG")) -> count_mags
#Genes
CountFeature(feature_c = "../data/clean/geneCover.txt",
           cutoff_pres = 90,
           method = "Assembly",
           mapfile = "../data/clean/mapfile.txt")  %>%
  mutate(feature = paste("Gene")) -> count_genes
#Species
CountFeature(feature_c = "../data/clean/speciesAbund.txt",
           cutoff_pres = 0.001,
           method = "Reads",
           mapfile = "../data/clean/mapfile.txt") %>%
  mutate(feature = paste("Species")) -> count_species

count_bins %>%
  bind_rows(count_mags) %>%
  bind_rows(count_genes) %>%
  bind_rows(count_species) %>%
write.table("../data/clean/coloniz_counts.txt", sep = "\t",
            quote = F, row.names = F, col.names = T)

```









