---
title: "parse input files"
output: html_document
date: "2025-07-10"
---

Here we are reading all results and mapfiles. creating required input files
for functions and analyses

```{r}

library(tidyverse)

```

## Mapfile

```{r}

mapfile <- read.csv("../data/mapfile.txt", sep = "\t")
data <- read.csv("../data/read_counts.txt", 
                 header = F, sep = ":")
data %>%
  filter(grepl("_R1.fastq.gz", V1)) %>%
  mutate(Sample= gsub("_R1.fastq.gz", "", V1)) %>%
  rename(Reads=V2) %>%
  left_join(mapfile) %>%
  select(Sample, Reads, Group, Isolator, Cage, Category) %>%
write.table("../data/clean/mapfile.txt", sep = "\t", 
            row.names = F, quote = F)

```

## MAG data
bins quality and taxonomy
```{r}

bin.qualy <- read.csv("../Results/Assembly/bins_quality.txt", sep = "\t")
bin.taxa <- read.csv("../Results/Assembly/bins_taxonomy.txt", sep = "\t")

bin.qualy %>%
  select(BinLab, Completeness, Contamination) %>%
  left_join(bin.taxa %>%
            select(BinLab, Phylum:Species)) %>%
write.table("../data/clean/binQT.txt", sep = "\t", 
            row.names = F, quote = F)

```




