---
title: "parse input files"
output: html_document
date: "2025-07-10"
---

Here we are reading all results and mapfiles. creating required input files
for functions and analyses

```{r}

library(tidyverse)
library(ape) # for reading/writing tree files
#load functions
source('./functions.R')

```

## Mapfile
###########################################################
One of parentA sample has a very low depth of sequencing
this must be removed from analyses
filter(Sample != "A1E10M")

```{r}

mapfile <- read.csv("../data/mapfile.txt", sep = "\t")
data <- read.csv("../data/read_counts.txt", 
                 header = F, sep = ":")
data %>%
  filter(grepl("_R1.fastq.gz", V1)) %>%
  mutate(Sample= gsub("_R1.fastq.gz", "", V1)) %>%
  rename(Reads=V2) %>%
  left_join(mapfile) %>%
  select(Sample, Reads, Group, Isolator, Cage, Category) %>%
  # sample with low depth:
  filter(Sample != "A1E10M") %>%
write.table("../data/clean/mapfile.txt", sep = "\t", 
            row.names = F, quote = F)

```

## MAG data

bins quality and taxonomy
```{r}

bin.qualy <- read.csv("../Results/Assembly/bins_quality.txt", sep = "\t")
bin.taxa <- read.csv("../Results/Assembly/bins_taxonomy.txt", sep = "\t")

bin.qualy %>%
  select(BinLab, Bin.Id,Completeness, Contamination) %>%
  left_join(bin.taxa %>%
            select(BinLab, Phylum:Species)) %>%
  # make a clean MAG label:
  mutate(family_lab= case_when(
  str_detect(Family, "\\d") ~ paste0("o_", Order),
  TRUE ~ paste0(Family))) %>%
  mutate(genus_lab = case_when(
  Genus == "" ~ paste0("f_", family_lab),
  str_detect(Genus, "\\d") ~ paste0("f_", family_lab),
  TRUE ~ paste0("g_", Genus))) %>%
  mutate(species_lab= case_when(
    Species == "" ~ paste0(genus_lab),
    str_detect(Species, "CAG") ~ paste0(genus_lab),
    str_detect(Species, "UBA") ~ paste0(genus_lab),
    str_detect(Species, "COE") ~ paste0(genus_lab),
    str_detect(Species, "RC9") ~ paste0(genus_lab),
    TRUE ~ Species)) %>%
  mutate(label = paste(species_lab, 
                        gsub("_", "", Bin.Id),
                        sep = "_")) %>%
  select(-Bin.Id, -family_lab, -genus_lab, -species_lab) %>%
write.table("../data/clean/binQT.txt", sep = "\t", 
            row.names = F, quote = F)

```

bins 1x coverage data across all samples. this is rarefied reads to lowest number
```{r}

contigs <- read.csv("../Results/Assembly/contig_in_bins_anvi.txt", sep = "\t",
                    header = F)
colnames(contigs) <- c("contig", "BinLab")

cover_parser(path = "../Results/p_map_subset/contigs_1k",
             patt= "_perfect.cover") %>%
  # adding bin info to contigs
  mutate(Don= gsub("_pool__.*", "", contig)) %>%
  left_join(contigs) %>%
  # define contigs outside of bins
  mutate(BinLab= case_when(is.na(BinLab) ~ paste(Don,"UnBin", sep = "_"),
                           TRUE ~ paste(BinLab))) %>%
  # Convert columns with numeric characters to numeric type
  mutate(across(everything(), ~ if (all(
    grepl("^\\s*-?\\d+(\\.\\d+)?\\s*$", .,
          perl = TRUE))) as.numeric(.) else .)) %>%
  group_by(Sample, BinLab) %>%
  summarise(
    bin_len = sum(endpos),
    bin_covbases = sum(covbases),
    bin_numreads = sum(numreads),
    bin_coverage = bin_covbases / bin_len * 100,
    .groups = "drop"
  ) %>%
  mutate(Sample = gsub("sub_", "", Sample)) %>%
  mutate(coverage = round(bin_coverage, digits = 2)) %>%
  rename(feature="BinLab") %>%
  select(Sample, feature, coverage) %>%
  # sample with low depth:
  filter(Sample != "A1E10M") %>%
write.table("../data/clean/binCover.txt", sep = "\t", 
            row.names = F, quote = F)
```

MAGs 1x coverage data across all samples.

```{r}

binQT <- read.delim("../data/clean/binQT.txt", 
           header = T, stringsAsFactors = F)

binCover <- read.delim("../data/clean/binCover.txt",
                       header = T, stringsAsFactors = F)

MAGid <- binQT %>%
  filter(Completeness >= 50 & Contamination <= 10) %>%
  pull(BinLab) %>% unique()

binCover %>%
  filter(feature %in% MAGid) %>%
write.table("../data/clean/MAGCover.txt", sep = "\t", 
            row.names = F, quote = F)

MAGid %>%
write.table("../data/clean/MAGids.txt", sep = "\t", 
            row.names = F, quote = F, col.names = F)

```

# MAGs detect in at least one sample of pups(AB/BA/ABAB):

```{r}

feature_c = "../data/clean/MAGCover.txt"
cutoff_pres = 75
method = "Assembly"
mapfile = "../data/clean/mapfile.txt"
# DonorA
donorA_feature <- DonFeature(feature_c = feature_c, donor = "DonA",
                           cutoff_pres = cutoff_pres, method = method)
finalA_identity <- identColoniz(donor_feature = donorA_feature, 
                               cutoff_pres = cutoff_pres,
                               mapfile = mapfile) 
finalA_identity %>% distinct(feature) %>%
  mutate(MAGs = gsub("DonA_bin_", "bin.", feature)) %>%
  mutate(MAGs = gsub("$", ".fa", MAGs)) %>%
  select(MAGs) %>%
  write.table("../data/clean/MAGids_DonA.txt", 
            row.names = F, quote = F, col.names = F)
# DonorB
donorB_feature <- DonFeature(feature_c = feature_c, donor = "DonB",
                           cutoff_pres = cutoff_pres, method = method)
finalB_identity <- identColoniz(donor_feature = donorB_feature, 
                               cutoff_pres = cutoff_pres,
                               mapfile = mapfile) 
finalB_identity %>% distinct(feature) %>%
  mutate(MAGs = gsub("DonB_bin_", "bin.", feature)) %>%
  mutate(MAGs = gsub("$", ".fa", MAGs)) %>%
  select(MAGs) %>%
  write.table("../data/clean/MAGids_DonB.txt", 
            row.names = F, quote = F, col.names = F)

```

MAG tree files, seperated for DonA and DonB
GTDB-tk was ran for MAGs and the alignment files were used to make a tree
using fasttree

```{r}

# Read a tree fix the label and keep it in clean data
DonA.tree <- read.tree("../Results/Assembly/DonA.bac120.user_msa.tree")
DonA.tree$tip.label <- gsub("bin.", "DonA_bin_", DonA.tree$tip.label)
write.tree(DonA.tree, "../data/clean/MAGs_DonA.tree")

DonB.tree <- read.tree("../Results/Assembly/DonB.bac120.user_msa.tree")
DonB.tree$tip.label <- gsub("bin.", "DonB_bin_", DonB.tree$tip.label)
write.tree(DonB.tree, "../data/clean/MAGs_DonB.tree")

DonAB.tree <- read.tree("../Results/Assembly/DonA_and_DonB.bac120.user_msa.tree")
DonAB.tree$tip.label <- gsub("DonA.bin.", "DonA_bin_", DonAB.tree$tip.label)
DonAB.tree$tip.label <- gsub("DonB.bin.", "DonB_bin_", DonAB.tree$tip.label)
write.tree(DonAB.tree, "../data/clean/MAGs_DonA_and_B.tree")

```

a UPGMA tree for clustering samples based on MAGs coverage information

```{r}

read.csv("../data/clean/mapfile.txt", sep = "\t") %>%
  mutate(Group = gsub("Parent", "P", Group),
         Group = gsub("Donor", "", Group)) %>%
  group_by(Group) %>%
  mutate(SG= paste(Group, row_number(), sep = "-")) %>%
  ungroup() %>%
  mutate(SG= gsub("A-1", "A", SG),
         SG= gsub("B-1", "B", SG)) %>%
  select(Sample, SG) -> mapfile

tree <- read.tree("../data/clean/MAGs_DonA_and_B.tree")
unique(tree$tip.label) -> ids

table <- read.csv("../data/clean/MAGCover.txt", sep = "\t") %>%
  filter(feature %in% ids) %>%
  left_join(mapfile) %>% select(-Sample)

wide_table <- table %>%
  pivot_wider(names_from = feature, values_from = coverage, values_fill = 0) %>%
  column_to_rownames("SG")

#Calculate distance matrix and perform UPGMA clustering
dist_matrix <- dist(wide_table, method = "euclidean")
hc <- hclust(dist_matrix, method = "average")  # UPGMA uses 'average' linkage

#Convert to phylo object and plot
tree <- as.phylo(hc)
write.tree(tree, "../data/clean/Samples_DonA_and_B.tree")

```

annotation file for iTOL heatmap

```{r}

read.csv("../data/clean/mapfile.txt", sep = "\t") %>%
  mutate(Group = gsub("Parent", "P", Group),
         Group = gsub("Donor", "", Group)) %>%
  group_by(Group) %>%
  mutate(SG= paste(Group, row_number(), sep = "-")) %>%
  ungroup() %>%
  mutate(SG= gsub("A-1", "A", SG),
         SG= gsub("B-1", "B", SG)) %>%
  select(Sample, SG) -> mapfile

tree <- read.tree("../data/clean/MAGs_DonA_and_B.tree")
ids <- unique(tree$tip.label)

# Read and filter coverage table
table <- read.csv("../data/clean/MAGCover.txt", sep = "\t") %>%
  filter(feature %in% ids) %>%
  left_join(mapfile) %>% select(-Sample)

# Pivot to wide format
heatmap_data <- table %>%
  spread(SG, coverage, fill = "X") %>%
  arrange(feature)

# Extract sample names
samples <- colnames(heatmap_data)[-1]  # exclude 'feature'

# Read and clean the Newick tree
upgma_tree <- readLines("../data/clean/Samples_DonA_and_B.tree") %>%
  paste(collapse = "") %>%
  str_replace_all("\\s+", "")  # remove spaces and newlines
# Create header with TAB separator
header <- c(
  "DATASET_HEATMAP",
  "SEPARATOR TAB",
  "DATASET_LABEL\texample_heatmap",
  "COLOR\t#ff0000",
  paste("FIELD_LABELS", paste(samples, collapse = "\t"), sep = "\t"),
  paste("FIELD_TREE", upgma_tree, sep = "\t"),
  "DATA"
)

# Create data section (tab-separated)
data_lines <- heatmap_data %>%
  mutate(across(-feature, ~ ifelse(is.na(.), "X", as.character(.)))) %>%
  rowwise() %>%
  mutate(line = paste(c_across(everything()), collapse = "\t")) %>%
  pull(line)

# Write to file
writeLines(c(header, data_lines),
           "../data/clean/itol_heatmap_annotation_tab.txt",
           useBytes = TRUE)

```


## Gene data

gff file
```{r}

gff_cols <- c("seqname","source","feature","start","end","score","strand",
              "frame","attributes")
read.delim("../Results/p_map_subset/Donor_db.gff",
           header=F, comment.char="#", ) -> gff
colnames(gff) <- gff_cols

# removing sequence info from the end of the gff file
gff %>%
  mutate(ID= gsub(";.*", "", attributes)) %>%
  mutate(gene_id=gsub("ID=", "", ID)) %>%
  # annotated regions that have only contig info in attributes. 
  # I don't need them
  filter(feature!= "region") %>%
  filter(feature == "CDS")-> gff
write.table(gff, "../data/clean/gene_db.txt", sep = "\t", 
            row.names = F, quote = F)


```

Gene 1x coverage across all samples. reads rarefied to the lowest depth.
```{r}

gff %>%
  mutate(donor=gsub("_pool__.*", "", seqname)) %>%
  mutate(feature= paste(donor, gene_id, sep = "__")) %>%
  select(gene_id, feature) -> gene.ids

cover_parser(path = "../Results/p_map_subset/genes",
             patt= "_perfect.cover") %>%
  dplyr::rename(gene_id = "contig") %>%
  # genes must be >=100bp
  filter(endpos >= 100) %>%
  # genes that were identified as CDS in gff file
  left_join(gene.ids) %>%
  filter(!is.na(feature)) %>%
  select(Sample, feature, coverage) %>%
   # sample with low depth:
  filter(Sample != "A1E10M") %>%
write.table("../data/clean/geneCover.txt", sep = "\t", 
            row.names = F, quote = F)

```

## Species data

Relative abundance of species across sample using metaphlan4
database used:#mpa_vJan21_CHOCOPhlAnSGB_202103

```{r}

read.csv("../Results/Reads/merged_metaphlan4.txt", 
         skip = 1, sep = "\t") %>%
  filter(str_detect(clade_name, "\\|t__")) %>%
  gather(Sample, coverage, -clade_name) %>%
  mutate(Sample = gsub(".mp.profile", "", Sample)) %>%
  rename(feature = "clade_name") %>%
  select(Sample, feature, coverage) %>%
  # sample with low depth:
  filter(Sample != "A1E10M") %>%
write.table("../data/clean/speciesAbund.txt", sep = "\t", 
            row.names = F, quote = F)

```

## StrainPhlan data
a list of species (SGB) with min abundance in donor and parent samples for
StrainPhlan analyses
```{r}

spAbund <- read.delim("../data/clean/speciesAbund.txt",
                      header = T, stringsAsFactors = F)
map <- read.delim("../data/clean/mapfile.txt", header = T, stringsAsFactors = F)

# fist, it must be present in donor samples
cut_off=0.001
spAbund %>%
  left_join(map) %>%
  filter(Group %in% c("DonorA", "DonorB")) %>%
  filter(coverage >= cut_off) %>%
  select(Sample, feature) %>%
  distinct() %>% pull(feature) -> DonSp

spAbund %>%
  filter(feature %in% DonSp) %>%
  left_join(map) %>%
  filter(Group %in% c("ParentA", "ParentB")) %>%
  select(Sample, feature, coverage) %>%
  spread(Sample, coverage) %>%
# Filter features with coverage > cut_off in at least 5 Parent samples
  rowwise() %>%
  mutate(above_cutoff = sum(c_across(-feature) > cut_off, na.rm = TRUE)) %>%
  ungroup() %>%
  filter(above_cutoff >= 5) %>%
  select(-above_cutoff) %>%
  mutate(SGB = gsub(".*\\|t__", "t__", feature)) %>%
  select(SGB, feature) -> SGBs

write.table(SGBs, "../data/clean/DonAndParentSp_list.txt", sep = "\t",
            quote = F, row.names = F)
SGBs %>%
  pull(SGB) %>%
write.table("../data/clean/DonAndParentSp_SGBs.txt", sep = "\t",
            quote = F, row.names = F, col.names = F)


```

importing strainphlan distances for the SGB of interest.note that some of these
were failed because they were not present in enough samples, with enough 
coverage. we used the authors suggested cut-off for these values and lower than
the default settings. 

```{r}

map <- "../data/clean/mapfile.txt"
precom_thresh = "../Results/Reads/strainphlan_threshold/VallesColomerM_2022_Jan21_thresholds.tsv"
path = "../Results/Reads/strainphlan"
patt = "_nGD.tsv"

data.frame(marker.id = paste(dir(path, pattern = patt))) %>%
  mutate(file_contents = map(marker.id, ~ read_tsv(
      file.path(path, .),
      col_names = F))) %>%
  unnest() %>%
  mutate(marker.id = gsub(patt, "", marker.id)) -> pwdis

length(unique(pwdis$marker.id))

pwdis %>%
  # only samples compared to Wild116, and X608:
  filter(X2 %in% c("X608", "Wild116")) %>%
  mutate(pairwise= paste(X1, X2, sep = "-")) %>%
  mutate(
         pairwise= paste(marker.id, pairwise, sep = "-"),
         distance= X3,
         Sample= paste(X1),
         SGB= marker.id,
         Donor= paste(X2)) %>%
  select(pairwise, SGB, Sample, distance, Donor) -> don_pwdis

```


counts of colonized features at different levels. one table of all features.

```{r}
#Bins
CountFeature(feature_c = "../data/clean/binCover.txt",
            cutoff_pres = 75,
            method = "Assembly",
            mapfile = "../data/clean/mapfile.txt") %>%
  mutate(feature = paste("Bin")) -> count_bins
#MAGs
CountFeature(feature_c = "../data/clean/MAGCover.txt",
            cutoff_pres = 75,
            method = "Assembly",
            mapfile = "../data/clean/mapfile.txt") %>%
  mutate(feature = paste("MAG")) -> count_mags
#Genes
CountFeature(feature_c = "../data/clean/geneCover.txt",
           cutoff_pres = 90,
           method = "Assembly",
           mapfile = "../data/clean/mapfile.txt")  %>%
  mutate(feature = paste("Gene")) -> count_genes
#Species
CountFeature(feature_c = "../data/clean/speciesAbund.txt",
           cutoff_pres = 0.001,
           method = "Reads",
           mapfile = "../data/clean/mapfile.txt") %>%
  mutate(feature = paste("Species")) -> count_species

count_bins %>%
  bind_rows(count_mags) %>%
  bind_rows(count_genes) %>%
  bind_rows(count_species) %>%
write.table("../data/clean/coloniz_counts.txt", sep = "\t",
            quote = F, row.names = F, col.names = T)

```

the size of database for each feature

```{r}

#Bins
DonFeature(feature_c = "../data/clean/binCover.txt", 
           cutoff_pres = 75, 
           donor = "DonA", method = "Assembly") %>%
  distinct(feature) %>%
  mutate(Donor = paste0("DonorA"),
         f.name= paste0("Bin")) -> Afeature_bins

DonFeature(feature_c = "../data/clean/binCover.txt", 
           cutoff_pres = 75, 
           donor = "DonB", method = "Assembly") %>%
  distinct(feature) %>%
  mutate(Donor = paste0("DonorB"),
         f.name= paste0("Bin")) -> Bfeature_bins
#MAGs
DonFeature(feature_c = "../data/clean/MAGCover.txt", 
           cutoff_pres = 75, 
           donor = "DonA", method = "Assembly") %>%
  distinct(feature) %>%
  mutate(Donor = paste0("DonorA"),
         f.name= paste0("MAG")) -> Afeature_mags
DonFeature(feature_c = "../data/clean/MAGCover.txt", 
           cutoff_pres = 75, 
           donor = "DonB", method = "Assembly") %>%
  distinct(feature) %>%
  mutate(Donor = paste0("DonorB"),
         f.name= paste0("MAG")) -> Bfeature_mags
#Genes
DonFeature(feature_c = "../data/clean/geneCover.txt", 
           cutoff_pres = 90, 
           donor = "DonA", method = "Assembly") %>%
  distinct(feature) %>%
  mutate(Donor = paste0("DonorA"),
         f.name= paste0("Gene")) -> Afeature_genes
DonFeature(feature_c = "../data/clean/geneCover.txt", 
           cutoff_pres = 90, 
           donor = "DonB", method = "Assembly") %>%
  distinct(feature) %>%
  mutate(Donor = paste0("DonorB"),
         f.name= paste0("Gene")) -> Bfeature_genes
#Species
DonFeature(feature_c = "../data/clean/speciesAbund.txt",
           cutoff_pres = 0.001,
           donor = "DonA", method = "Reads") %>%
  distinct(feature) %>%
  mutate(Donor = paste0("DonorA"),
         f.name= paste0("Species")) -> Afeature_species
DonFeature(feature_c = "../data/clean/speciesAbund.txt",
           cutoff_pres = 0.001,
           donor = "DonB", method = "Reads") %>%
   distinct(feature) %>%
   mutate(Donor = paste0("DonorB"),
         f.name= paste0("Species")) -> Bfeature_species

Afeature_bins %>%
  bind_rows(Bfeature_bins,
            Afeature_mags,
            Bfeature_mags,
            Afeature_genes,
            Bfeature_genes,
            Afeature_species,
            Bfeature_species) -> table

table %>%
  group_by(Donor, f.name) %>%
  summarise(feature_size= n()) %>%
  rename(feature = "f.name") %>%
  write.csv("../data/clean/feature_size.txt", sep = "\t",
            quote = F, row.names = F, col.names = T)

```

input data for stats models

```{r}

#load mapfile
mapfile <- read.csv("../data/clean/mapfile.txt", sep = "\t")
mapfile %>%
  distinct(Sample, Cage) -> mapfile

# load the data
coloniz <- read.csv("../data/clean/coloniz_counts.txt",
                    sep = "\t")
coloniz <- coloniz %>%
  filter(Group %in% c("AB", "BA", "ABAB")) %>%
  left_join(mapfile)

#load the feature.size
feature_size <- read.csv("../data/clean/feature_size.txt")

coloniz %>%
  mutate(order = case_when(
    Donor == "DonorA" & Group == "AB" ~ paste0("first"),
    Donor == "DonorA" & Group == "BA" ~ paste0("second"),
    Donor == "DonorB" & Group == "BA" ~ paste0("first"),
    Donor == "DonorB" & Group == "AB" ~ paste0("second"),
    Group == "ABAB" ~ paste("mixed"),
    TRUE ~ paste0("none")
    )) %>%
  mutate(log.coloniz = log(n_coloniz)) %>%
  left_join(feature_size)-> colniz

# Munge data
coloniz = (colniz
           %>% mutate(order = factor(order, levels = c('mixed', 'second', 'first')),
                     feature = factor(feature, levels = c('Bin', 'MAG', 'Species', 'Gene')))
           %>% group_by(feature)
           %>% rename(log_count = log.coloniz,
                      raw_count = n_coloniz)
           %>% mutate(z_count = scale(raw_count)[,1],
                      prop_count = raw_count/feature_size,
                      zprop_count = scale(prop_count)[,1])
           
           %>% ungroup())


write.table(coloniz, "../data/clean/coloniz_model.txt", sep = "\t",
            quote = F, row.names = F, col.names = T)


```

input data for visualization of counts

```{r}

#load mapfile
mapfile <- read.csv("../data/clean/mapfile.txt", sep = "\t")

# load the data
coloniz <- read.csv("../data/clean/coloniz_counts.txt",
                    sep = "\t")
#load the feature.size
feature_size <- read.csv("../data/clean/feature_size.txt")

coloniz %>%
  mutate(order = case_when(
    # pups
    Donor == "DonorA" & Group == "AB" ~ paste0("first"),
    Donor == "DonorA" & Group == "BA" ~ paste0("second"),
    Donor == "DonorB" & Group == "BA" ~ paste0("first"),
    Donor == "DonorB" & Group == "AB" ~ paste0("second"),
    Group == "ABAB" ~ paste("mixed"),
    # parents
    Donor == "DonorA" & Group == "ParentA" ~ paste0("parent"),
    Donor == "DonorA" & Group == "ParentB" ~ paste0("non-parent"),
    Donor == "DonorB" & Group == "ParentA" ~ paste0("non-parent"),
    Donor == "DonorB" & Group == "ParentB" ~ paste0("parent"),
    Group == "ParentAB" ~ paste("mixed-parent"),
    # Donors
    Donor == "DonorA" & Group == "DonorA" ~ paste0("donor"),
    Donor == "DonorA" & Group == "DonorB" ~ paste0("non-donor"),
    Donor == "DonorB" & Group == "DonorA" ~ paste0("non-donor"),
    Donor == "DonorB" & Group == "DonorB" ~ paste0("donor"),
    TRUE ~ paste0("none")
    )) %>%
  mutate(order = factor(order, levels = c("donor", "non-donor",
                                          "parent", "mixed-parent", 
                                          "non-parent",
                                          "first", "mixed", "second"))) -> coloniz

write.table(coloniz, "../data/clean/coloniz_visual.txt", sep = "\t",
            quote = F, row.names = F, col.names = T)



```

## shannon diversity

```{r}

read.csv(
"../Results/Reads/diversity_analysis/merged_metaphlan4_shannon.tsv",
sep = "\t") %>%
  rownames_to_column("Sample") %>%
  mutate(Sample = gsub(".mp.profile", "", Sample)) %>%
  # sample with low depth:
  filter(Sample != "A1E10M") %>%
  rename(Shannon = "diversity_shannon") %>%
write.table("../data/clean/diversity_shannon.txt", sep = "\t", 
            row.names = F, quote = F)

```

## bray-curtis distance

```{r}

read.csv("../data/clean/mapfile.txt", sep = "\t") %>%
  select(Sample, Group)-> mapfile
read.csv("../Results/Reads/diversity_analysis/merged_metaphlan4_bray-curtis.tsv",
         sep = "\t") %>%
  rownames_to_column("Sample_1") %>%
  gather(Sample_2, value, -Sample_1) %>%
  filter(as.character(Sample_1) != as.character(Sample_2)) %>%
  mutate_if(is.factor, as.character) %>%
  mutate(Sample_1 = gsub(".mp.profile", "", Sample_1),
         Sample_2 = gsub(".mp.profile", "", Sample_2))-> dis.tbl

# two mapfile for each variable in matrix
mapfile -> V1.tbl
colnames(V1.tbl) <- paste(colnames(V1.tbl), "1", sep = "_")
mapfile-> V2.tbl
colnames(V2.tbl) <- paste(colnames(V2.tbl), "2", sep = "_")

#merging all info together
dis.tbl %>%
  left_join(V1.tbl) %>%
  left_join(V2.tbl) %>%
  # comparing all samples to donor's
  filter(Sample_1 %in% c("X608", "Wild116")) %>%
  filter(!Sample_2 %in% c("X608", "Wild116")) %>%
  # One of parent A sample has a very low depth of sequencing
  # removing this sample from the run
  filter(Sample_2 != "A1E10M") %>%
write.table("../data/clean/paired_bray.txt", sep = "\t", 
            row.names = F, quote = F)


```

## contig information

```{r}

read.csv("../Results/Assembly/Wild116_contigs.list",
         header = FALSE) %>%
  rename(NAME = "V1") %>%
  mutate(Sample = paste("Wild116")) -> Wild116

read.csv("../Results/Assembly/X608_contigs.list",
         header = FALSE) %>%
  rename(NAME = "V1") %>%
  mutate(Sample = paste("X608")) -> X608

cutoff=1000
Wild116 %>%
  bind_rows(X608) %>%
  mutate(LENGTH = gsub(".*_length_", "", NAME)) %>%
  mutate(LENGTH = gsub("_cov_.*", "", LENGTH)) %>%
  # select those that are above 1K
  filter(as.numeric(LENGTH) >= cutoff) %>%
  group_by(Sample) %>%
  mutate(Cumulative= cumsum(as.numeric(LENGTH))) %>%
  mutate(Contigid = row_number()) %>%
  mutate(Cumulative = Cumulative / 1000000) %>%
  mutate(Contigid2 = Contigid / 1000) %>%
  mutate(Sample = case_when(Sample == "Wild116" ~ paste("DonorB"),
                            Sample == "X608" ~ paste("DonorA"),
                            TRUE ~ paste("idk"))) %>%
  mutate(Method = paste("Single")) %>%
  select(Sample:Method, -LENGTH)-> single_assem

fai_head <- c("NAME", "LENGTH", "OFFSET",
              "LINE_BLEN", "LINE_FBLEN")

read.csv("../Results/Assembly/DonA_and_B_contigs_1k.fa.fai", 
         sep = "\t", header  = F) -> contig

colnames(contig) <- fai_head
contig %>%
  mutate(Sample = gsub("_pool__.*", "", NAME)) %>%
  group_by(Sample) %>%
  mutate(Cumulative= cumsum(as.numeric(LENGTH))) %>%
  group_by(Sample) %>%
  mutate(Contigid = row_number()) %>%
  mutate(Cumulative = Cumulative / 1000000) %>%
  mutate(Contigid2 = Contigid / 1000) %>%
   mutate(Sample = case_when(Sample == "DonA" ~ paste("DonorA"),
                            Sample == "DonB" ~ paste("DonorB"),
                            TRUE ~ paste("idk"))) %>%
  mutate(Method = paste("Co-assembly")) %>%
  select(Sample:Method)-> co_assem

single_assem %>%
  bind_rows(co_assem) %>%
write.table("../data/clean/assembly_quality.txt", sep = "\t", 
            row.names = F, quote = F)

```

## generate iTOL annotation file

```{r}


BinQT <- read.csv("../data/clean/binQT.txt", sep = "\t")

# Prepare the annotation lines
header <- c(
  "DATASET_LABELS",
  "SEPARATOR TAB",
  "DATASET_NAME\tSpecies Labels",
  "FIELD_LABELS\tSpecies",
  "FIELD_COLORS\t#ff0000",
  "DATA"
)

# Select and rename columns
tip_labels <- BinQT %>%
  select(BinLab, label) %>%
  rename(tip_id = BinLab, species_label = label)

# Create label lines
label_lines <- paste(tip_labels$tip_id, tip_labels$species_label, sep = "\t")

# Combine header and data
itol_annotation <- c(header, label_lines)

# Write to file
writeLines(itol_annotation, "../data/clean/itol_labels.txt")


```


```{r}

BinQT <- read.csv("../data/clean/binQT.txt", sep = "\t")

# Define phylum colors
phylum_col <- c(
  "Firmicutes_A" = "#6a3d9a",
  "Actinobacteriota" = "#33a02c",
  "Bacteroidota" = "#1f78b4",
  "Patescibacteria" = "#ff7f00",
  "Firmicutes" = "#b2df8a",
  "Desulfobacterota" = "#fdbf6f",
  "Firmicutes_B" = "#ffff99",
  "Deferribacterota" = "#a6cee3",
  "Proteobacteria" = "#e31a1c",
  "Campylobacterota" = "#f781bf",
  "NA" = "#808080"
)

# Prepare header
header <- c(
  "DATASET_STYLE",
  "#Style datasets allow the customization of branch and leaf label colors and styles.",
  "SEPARATOR COMMA",
  "DATASET_LABEL,Phylum Branch Colors",
  "COLOR,#000000",
  "DATA"
)

# Create style lines for each tip with branch coloring
style_lines <- BinQT %>%
  mutate(branch_color = phylum_col[as.character(Phylum)]) %>%
  mutate(branch_color = ifelse(is.na(branch_color), "#000000", branch_color)) %>%
  mutate(line = paste(BinLab, "branch", "node", branch_color, "2", "normal", sep = ",")) %>%
  pull(line)

# Combine header and data
itol_branch_style <- c(header, style_lines)

# Write to file
writeLines(itol_branch_style, "../data/clean/itol_phylum_branch_colors.txt")


```


```{r}

BinQT <- read.csv("../data/clean/binQT.txt", sep = "\t")

# Define phylum colors
phylum_col <- c(
  "Firmicutes_A" = "#6a3d9a",
  "Actinobacteriota" = "#33a02c",
  "Bacteroidota" = "#1f78b4",
  "Patescibacteria" = "#ff7f00",
  "Firmicutes" = "#b2df8a",
  "Desulfobacterota" = "#fdbf6f",
  "Firmicutes_B" = "#ffff99",
  "Deferribacterota" = "#a6cee3",
  "Proteobacteria" = "#e31a1c",
  "Campylobacterota" = "#f781bf",
  "NA" = "#808080"
)

# Prepare header for iTOL color strip file
header <- c(
  "DATASET_COLORSTRIP",
  "SEPARATOR COMMA",
  "DATASET_LABEL,Phylum Strip Colors",
  "COLOR,#000000",
  "STRIP_WIDTH,25",
  "MARGIN,5",
  "SHOW_INTERNAL,0",
  "DATA"
)

# Create strip lines for each tip
strip_lines <- BinQT %>%
  mutate(strip_color = phylum_col[as.character(Phylum)]) %>%
  mutate(strip_color = ifelse(is.na(strip_color), "#000000", strip_color)) %>%
  mutate(line = paste(BinLab, strip_color, Phylum, sep = ",")) %>%
  pull(line)

# Combine header and data
itol_strip <- c(header, strip_lines)

# Write to file
writeLines(itol_strip, "../data/clean/itol_phylum_colorstrip.txt")

```

## commonly colonized genes

```{r}

mapfile <- read.csv("../data/clean/mapfile.txt", sep = "\t")
mapfile %>%
  filter(Group %in% c("AB", "BA", "ABAB")) %>%
  pull(Sample) -> Spup

cutoff_pres <- 90
g <- read.csv("../data/clean/geneCover.txt", sep = "\t")

# Join with group info
g <- g %>%
  filter(Sample %in% Spup) %>%
  left_join(mapfile %>% select(Sample, Group), by = "Sample")

# Filter for coverage >= cutoff
g_filtered <- g %>%
  filter(coverage >= cutoff_pres)

# Count how many samples per group each gene appears in
gene_group_counts <- g_filtered %>%
  group_by(Group, feature) %>%
  summarise(n_detected = n_distinct(Sample), .groups = "drop")

# Total samples per group
samples_per_group <- mapfile %>%
  filter(Sample %in% Spup) %>%
  group_by(Group) %>%
  summarise(n_total = n_distinct(Sample), .groups = "drop")

# Join and filter for genes detected in all of samples per group (100%)
common_genes <- gene_group_counts %>%
  left_join(samples_per_group, by = "Group") %>%
  filter(n_detected / n_total >= 1) %>%
  mutate(temp=feature) %>%
  separate(temp, c("Donor", "gene_id"), sep = "__") %>%
  mutate(order = case_when(
    # pups
    Donor == "DonA" & Group == "AB" ~ paste0("first"),
    Donor == "DonA" & Group == "BA" ~ paste0("second"),
    Donor == "DonB" & Group == "BA" ~ paste0("first"),
    Donor == "DonB" & Group == "AB" ~ paste0("second"),
    Group == "ABAB" ~ paste("mixed"),
    TRUE ~ paste("none")))

common_genes %>%
  select(feature, Group) %>% mutate(n = paste(1)) %>%
  spread(Group, n) %>%
  mutate(ptype= case_when(AB == 1 & BA == 1 & ABAB == 1 ~ paste("All"),
                         AB == 1 & is.na(BA) & is.na(ABAB) ~ paste("AB"),
                         AB == 1 & BA == 1 & is.na(ABAB) ~ paste("AB-BA"),
                         AB == 1 & is.na(BA) & ABAB == 1 ~ paste("AB-ABAB"),
                         is.na(AB) & BA == 1 & is.na(ABAB) ~ paste("BA"),
                         is.na(AB) & BA == 1 & ABAB == 1 ~ paste("BA-ABAB"),
                         is.na(AB) & is.na(BA) & ABAB == 1 ~ paste("ABAB"),
                         TRUE ~ paste("other"))) %>%
  select(feature, ptype) -> common_genes_overlap

common_genes %>%
  left_join(common_genes_overlap) -> common_genes

common_genes %>%
  select(feature) %>% distinct() %>%
  write.table("../data/clean/commonly_colonized_genes.txt", 
            sep = "\t", row.names = F, col.names = F, quote = F)

```


```{r}

common_genes %>%
  select(Donor, gene_id, order) %>%
  group_by(Donor, order) %>% summarise(n = n()) %>%
  ggplot(aes(order, n, color = Donor)) +
  geom_point() + geom_line(aes(group = Donor)) +
  theme_bw() +
  ylab("commonly colonized genes")
  


```


```{r}

ggplot(common_genes, aes(gene_id, Group, fill=Donor)) +
  geom_tile() +
  facet_grid(~ptype, scales = "free", space = "free") +
  theme_classic() +
  theme(axis.text.x = element_blank(),
        strip.text.x = element_text(angle = 90))

```


